[WIP] Add support to filter channels by dropdown in stream settings.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First 3 commits are refactoring commits.
- Next commit is to add folder filter.
- (To be squashed) Last commit is to add chevron icon to folder filter icon.

Fixes: <!-- Issue link, or clear description.--> #35779

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
Added code to show only folders in which there are channels accessible to user. There is a problem for guests in this case, as guest have access to unsubscribed web-public streams but those are not shown in UI. So, the folder is shown even when there are no channels.

Do we want that or do we want to handle guest case separately by showing only folders which contain guest's subscribed channels?

Also what about "Uncategorized" option? Do we want to show it only when there is at least one accessible channel which is not in any folder? Currently the PR does this.

alya commented:
Yeah, I think the icon change is still good. I'll test again once we've got an updated icon from @karlstolley .

alya commented:
1. We now need to horizontally align the filter box with the channel icons.
2. I'm not liking how bold the bold looks in the dropdown, especially with the yellow. Can we make it less bold? Probably our dropdowns woulld benefit from that change in general.

<img width="249" height="146" alt="Screenshot 2025-09-03 at 14 27 34" src="https://github.com/user-attachments/assets/6630fb97-c2bc-4b87-867d-9e0ff727fdd0" />

3. Something is wrong with how the keyboard selection box looks.
<img width="98" height="57" alt="Screenshot 2025-09-03 at 14 28 34" src="https://github.com/user-attachments/assets/2e12c39a-6048-4c43-b1c3-bdec40b931e0" />
4. I'd like more of a delay on the info tooltip. It should only pop up if I pause on the folder dropdown, not if I'm just trying to click it.

sahil839 commented:
Updated the PR.
- Squashed the commits for testing different icons. Now we show the new icon created by Karl with chevron when showing the default state and https://lucide.dev/icons/folder-search when a filter is selected.
- Updated the alignment as discussed on CZO.
- Added delay when showing tooltip.

alya commented:
I guess the button selection box is as designed, except that we should make it nice and square. I don't know if there are CSS changes top be made more generally to ensure this? Here's what we have in /devtools/buttons, based on the [design in Figma](https://www.figma.com/design/msWyAJ8cnMHgOMPxi7BUvA/Zulip-Web-UI-kit?node-id=8-1681&t=bmpReOb9lf8S4fJ1-0).

<img width="168" height="103" alt="Screenshot 2025-09-04 at 12 57 25" src="https://github.com/user-attachments/assets/9ea61301-1283-4079-813c-2284b400c62a" />
<img width="119" height="52" alt="Screenshot 2025-09-04 at 12 58 08" src="https://github.com/user-attachments/assets/74c2b73e-bd46-40c4-8ebc-651fedbd8bd4" />

There's an alternate option, which we might want to try to see if it looks better:
<img width="142" height="97" alt="Screenshot 2025-09-04 at 12 57 29" src="https://github.com/user-attachments/assets/c894f393-d6af-469f-b114-09a7c01500f9" />
<img width="123" height="62" alt="Screenshot 2025-09-04 at 12 58 12" src="https://github.com/user-attachments/assets/a37ff49a-ad5c-4e95-a629-163a9bbe130d" />

alya commented:
600 looks fine to me in the inbox dropdown. I think we should use the background highlighting on the selected option that we have in the inbox in this dropdown too, which will make it a lot easier to spot.

sahil839 commented:
I have added the commit to add background highliting on the selected option. But there is some difference in how the items are focussed during keyboard navigation. It depends on whether the dropdown has search or not.

So, for the folder filter background highlighting is done when navigating using keyboard but for other dropdowns like inbox filter, archived status filter we just show a blue outline.

It will require some careful refactor for changing the folder filter to have outline during keyboard navigation so I have not done it yet. We can handle that later if the current version seems fine, otherwise I can fix that.

![Peek 2025-09-08 13-22](https://github.com/user-attachments/assets/806b4b81-778d-490b-8a3f-f75596858101)

timabbott commented:
I opened #36909 to integrate the preparatory refactoring commits.

timabbott commented on web/src/stream_settings_components.ts:
```diff
@@ -315,6 +332,21 @@ export function archived_status_filter_includes_channel(sub: StreamSubscription)
+            return false;
+        }
+    }
+    return true;
```
Ended up refactoring this to be more readable like this:

``` diff
diff --git a/web/src/stream_settings_components.ts b/web/src/stream_settings_components.ts
index 6ed1a93529..09f685b004 100644
--- a/web/src/stream_settings_components.ts
+++ b/web/src/stream_settings_components.ts
@@ -335,16 +335,14 @@ export function archived_status_filter_includes_channel(sub: StreamSubscription)
 export function folder_filter_includes_channel(sub: StreamSubscription): boolean {
     const filters = stream_settings_data.FOLDER_FILTERS;
     const filter_value = get_folder_filter_dropdown_value();
-    if (filter_value !== filters.ANY_FOLDER_DROPDOWN_OPTION) {
-        if (filter_value === filters.UNCATEGORIZED_DROPDOWN_OPTION) {
-            if (sub.folder_id !== null) {
-                return false;
-            }
-        } else if (filter_value !== sub.folder_id) {
-            return false;
-        }
+    if (filter_value === filters.ANY_FOLDER_DROPDOWN_OPTION) {
+        return true;
     }
-    return true;
+
+    if (filter_value === filters.UNCATEGORIZED_DROPDOWN_OPTION) {
+        return sub.folder_id === null;
+    }
+    return filter_value === sub.folder_id;
 }
 
 export function set_up_folder_dropdown_widget(sub?: StreamSubscription): DropdownWidget {
```
